### ISSUE TITLE
Override JpaContentTemplate

### ISSUE THREAD

**Issue by brunnels** [paulcwarren/spring-content#19]
I'm trying to use a sql server backend which doesn't work with the sql in the JpaContentTemplate.  What is the best way to override the bean that is created in JpaStoresRegistrar?  Would be great if spring boot autoconfig looked for an existing bean.  Setting my bean as @Primary and giving it the same name is also not working.

**Comment by paulcwarren** [comment]
Hi @brunnels.  Thanks for raising the issue.  That doesn't surprise me.  The current JPA module is a little bit of a POC.  However, we are wrapping up a major revision that takes the same approach as Spring Batch with platform-specific .SQL files and Spring Batch auto-configuration that picks and runs the right set of .SQL files.  

This sounds pretty close/almost exactly what u need.  Would u be willing to test/contribute .SQL files for SQL Server?

**Comment by paulcwarren** [comment]
Hi @brunnels.  Thanks for raising the issue.  That doesn't surprise me.  The current JPA module is a little bit of a POC.  However, we are wrapping up a major revision that takes the same approach as Spring Batch with platform-specific .SQL files and Spring Batch auto-configuration that picks and runs the right set of .SQL files.  

This sounds pretty close/almost exactly what u need.  Would u be willing to test/contribute .SQL files for SQL Server?

**Comment by brunnels** [comment]
Sure.  Just point me at the branch and I can send a merge request.

**Comment by paulcwarren** [comment]
Great.  Give me a day or two to pull it together and push the latest of the feature branch and I'll ping you here.  Thanks

**Comment by paulcwarren** [comment]
Hey @brunnels I have checked in a "work in progress" on the wip/jpa-resource branch.  It isn't particularly clean (in terms of test-driven) but it has been manually tested against mysql and postgres.   

On this branch the ways the JPA module is intended to work the same way as Spring Batch (another Spring library that requires schema initialization).  When using Spring Content for JPA the developer will have to supply the DataSource bean (as he always has) and a DataSourceInitializer bean to initialize the given DataSource with the relevant Spring Content schema.  in this case the DataSouceInitializer would look like this:-

```
    @Value("/org/springframework/content/jpa/schema-drop-postgresql.sql")
    private Resource dropReopsitoryTables;

    @Value("/org/springframework/content/jpa/schema-postgresql.sql")
    private Resource dataReopsitorySchema;

    @Bean
    public DataSource dataSource() {
        ...
        return ds;
    }

    @Bean
    DataSourceInitializer datasourceInitializer() {
        ResourceDatabasePopulator databasePopulator =
                new ResourceDatabasePopulator();

        databasePopulator.addScript(dropReopsitoryTables);
        databasePopulator.addScript(dataReopsitorySchema);
        databasePopulator.setIgnoreFailedDrops(true);

        DataSourceInitializer initializer = new DataSourceInitializer();
        initializer.setDataSource(dataSource());
        initializer.setDatabasePopulator(databasePopulator);

        return initializer;
    }
``` 

* entity transaction manager beans ommitted for brevity

I have (manually) tested this using the example/integration project spring-eg-content-jpa by hacking up the configuration to provide the new datasourecinitializer bean.  

Point to note that when using the boot starter for Spring Content for JPA this datasource initialization will eventually be auto-configured.  This is still to be done BTW so don't expect it to work.

I also refactored the central DefaultJPAStoreImpl class to use a new (Blob)Resource abstraction based on the standard Spring Resource interface (this is actually part of a broader experimental Store API piece of work).    This is where the relevant SQL will be executed as content is set, updated and unset (see AbstractBlobResource).  There is a possibility that the hardcoded SQL statements wont work on SQLServer and if that is the case we should first take a look at whether we have done anything non-standard in the SQL, or not.   Possibly the case.  But I am also finding that the various drivers don't support the full compliment of JDBC API and we *may* be forced to have database-specific blob resource implementations.  In a previous incantation of this changed I had actually paved the way for this so I left that code in just in case we need it.  See DelegatingBlobResourceLoader and AbstractBblobResource.  Feel free to use it if you feel that you need to.  However, if we don't end up using it then we can decide later on whether to leave it in, or rip it out.  This can of plug-in capability can often useful and can get someone of out trouble in a pinch.  But maybe it is more hassle than it is worth.

So I think you can use this to test against SQLServer by adding a the relevant .sql files and adding the DataSourceInitializer bean to your project then set, updating and unsettting content.  feel free to post back here with anything that you do find.

In the meantime, I will clean this all up, go find various free hosting options to perform accepting testing against and add the spring boot starter auto-configuration.

**Comment by paulcwarren** [comment]
Hi @brunnels I just committed changes to support sql server on wip/jpa-resource.  Our integration tests work but aren't committed yet).  Please feel free to give this branch a go but I will try and wrap this up and merge it across to master soonest whereby it will be available in the latest snapshot.

**Comment by brunnels** [comment]
Sorry for the delay in response.  I have a production release due end of
the month that's been taking all my time.  I'll give it a try asap and
report back.  Thank you for the quick turnaround.

**Comment by paulcwarren** [comment]
No problem.  I know how it is.  When you get a chance let me know how it goes.

**Comment by paulcwarren** [comment]
I rebased this work onto master this evening so it is available on the latest 0.0.10-SNAPSHOT.  Might make it a bit easier to consume for you.
